### PR TITLE
PDF - implement customDrawCell callback

### PR DESCRIPTION
**Pull request by novsstation**

### CODE REVIEW

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                const done = assert.async();
+                const doc = createMockPdfDoc();
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    pdfCell.customDrawCell = ({ rect, cell }) => {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: ({ rect, cell }) -> ({ jsPDFDocument, rect, cell })
\+ assert.strictEqual(jsPDFDocument, doc);

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                        assert.deepEqual(rect, { h: 18.4, w: 500, x: 10, y: 10 });
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: There should be correct values for each cell: let's use "if(pdfCell.text === 'f1') { pdfCell.customDrawCell = .." approach

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f2') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.addImage(logoBase64, 'JPEG', rect.x, rect.y, 16, 16);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f1') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.addImage(logoBase64, 'JPEG', rect.x, rect.y, 16, 16);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f2') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.textWithLink('Click here', rect.x, rect.y + rect.h - 1, { url: 'http://www.google.com' });
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f1') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.textWithLink('Click here', rect.x, rect.y + rect.h - 1, { url: 'http://www.google.com' });
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f2') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.rect(rect.x, rect.y, rect.w, rect.h, 'F');
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }, { caption: 'f2' }]
+                });
+
+                const customizeCell = ({ gridCell, pdfCell }) => {
+                    if(pdfCell.text === 'f1') {
+                        pdfCell.customDrawCell = ({ rect }) => {
+                            doc.rect(rect.x, rect.y, rect.w, rect.h, 'F');
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: It's not clear what we are testing here

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+                });
+            });
+
+            QUnit.test('customDrawCell check event arg', function(assert) {
+                const done = assert.async();
+                const doc = createMockPdfDoc();
+
+                const dataGrid = createDataGrid({
+                    width: 500,
+                    columns: [{ caption: 'f1' }]
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: \+ columns: [{ caption: 'f1', caption: 'f2' }]

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/exporter/jspdf/v3/draw_utils.js [hunk 10]
```diff
@@ -11,7 +11,13 @@ function round(value) {
 function drawCellsContent(doc, cellsArray, docStyles) {
     cellsArray.forEach(cell => {
         drawCellBackground(doc, cell);
-        drawCellText(doc, cell, docStyles);
+        if(isDefined(cell.customDrawCell)) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/v3/draw_utils.js]: drawCellBackground(doc, cell); - why it doesn't depend from customDrawCell ?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 1 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Yep, great idea. Fixed

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 2 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Yep, fixed. Now I check all available cells in this test

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 3 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: I wanted to test that the new functionality allows customers to draw any rects/images/links/etc in a cell.
But I agree, in fact, I got into checking the logic of the jsPDF.
After I added a checking of the doc argument, this test now looks redundant :)
Removed it.

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 4 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Same as before. Removed this test

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 5 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Same as before. Removed it

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 6 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Same as before. Removed it

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 7 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Same as before. Removed it

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 8 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Same as before. Removed it

File: testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js (hunk 9 above)
Comment (novsstation) [reply testing/tests/DevExpress.exporter/jspdfParts/v3/jspdf_v3.dataGrid.customDrawCell.tests.js]: Yep, fixed and now I checks all available cells

File: js/exporter/jspdf/v3/draw_utils.js (hunk 10 above)
Comment (novsstation) [reply js/exporter/jspdf/v3/draw_utils.js]: My fault. 
Now, I call the drawCellBackground only if customDrawCell is not specified

**Review by KuznetsovVN — DISMISSED** [review]

**Review by KuznetsovVN — DISMISSED** [review]

**Review by IgnatovDan — APPROVED** [review]

**Review by KuznetsovVN — APPROVED** [review]

**Pull request merged**